### DISCUSSION TITLE
Fetch caching/memoization doesn't seem to work in App Router when changing route

### DISCUSSION THREAD

**Discussion by cpotey** [vercel/next.js#56428] [Help]
### Summary

I have created a [couple of functions to fetch data](https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#using-react-cache-server-only-and-the-preload-pattern) from my API (listed below), I thought that everytime it would cache the response (as the function arguments haven't changed for my `fetchBanner` function) - but everytime the route changes the `fetchBanner` data is refetched again, even if though it's an identical function, with identical arguments as the previous route?

My assumption with how this cache/memoization would work is that when a cached function is called, it will reuse the same response as a previous one rather than refetch? - but I can see the endpoint being requested everytime the route changes for my `fetchBanner` request.

Though, if I go to a page: Page A (it fetches the data), then I navigate to another: Page B (it then fetches Page B's data), and then if I return to Page A, the data is cached - and doesn't do a new fetch request, so that's something at least... but I currently will have to do the `fetchBanner` request everytime I land on a new page even though nothing has changed!

Am I misunderstanding how the [cache/memoization](https://react.dev/reference/react/cache) should be working? It seems incredibly inefficient for this to be happening when the response/request is no different across each route change, the only change is the route. It's caching but caching seems tied to the route?

(Everything is working fine, navigating around pages and using the fetchPageData etc - it's the the fetchBanner that is really throwing me - I feel like I'm missing something?)

### Additional information

`./app/[...slug]/page.tsx`
```js
import { fetchPageData, fetchBanner } from '@utils/api/requests';

export default async function Page({
  params
}: {
  params: { slug: Array<string | undefined> };
}) {
  const data = await fetchPageData(params?.slug?.join('/'));
  const bannerData = await fetchBanner('default-marketing-footer'); // This gets called every time even though the argument is exactly the same on every route
 
  return (
    <main>
       <h1>{data.title}</h1>
    </main>
  );
}
```

`./app/(homepage)/page.tsx`
```js
import { fetchPageData, fetchBanner } from '@utils/api/requests';

export default async function Homepage() {
  const bannerData = await fetchBanner('default-marketing-footer'); 
 
  return (
    <main>
       <h1>Homepage</h1>
    </main>
  );
}
```

`./utils/api/requests.ts`
```js
import { cache } from 'react';

export const fetchPageData = cache(async (pageSlug: string): Promise<Page> => {
  const res = await fetch(
    `http://myapi.com/pages/with-path/${pageSlug !== '/' ? `${pageSlug}/` : ''}`,
    {
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
    }
  );

  if (!res.ok) {
    throw new Error('Failed to fetch page data');
  }
  return res.json() as Promise<Page>;
});

export const fetchBanner = cache(async (bannerSlug: string): Promise<Banner> => {
  const res = await fetch(`http://myapi.com/banners/?slug=${bannerSlug}`, {
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
  });

  if (!res.ok) {
    throw new Error('Failed to fetch banner data');
  }
  return res.json() as Promise<Banner>;
});
```

**Comment by mdomarus** [comment]
Have you figured it out?

**Reply by cpotey** [reply]
Nope, seems that this _would_  work if my fetch requests didn't use Cookies/custom headers (not included in the above example tbf) - this would use the default Next fetching. Adding headers/cookies to the fetch requests makes it dynamic and excludes it from Next's fetch caching. As far as I'm aware, this default Next fetch cache WOULD work through different routes - but I can't use that in my scenario.

Using React Cache instead basically seems to be doing memoization only, but only per route. Every time you change route it will refetch/cache/memoize, which is a pain (but if you return to a previous route, will use the old cached data) - but haven't got any other solution for now either.
